File: backends/dotnet/shared/Implementations/Middleware/RequestEnrichment/FingerprintBuilder.cs
// -----------------------------------------------------------------------
// <copyright file="FingerprintBuilder.cs" company="DCSV">
// Copyright (c) DCSV. All rights reserved.
// </copyright>
// -----------------------------------------------------------------------

namespace D2.Shared.RequestEnrichment;

using System.Security.Cryptography;
using System.Text;
using Microsoft.AspNetCore.Http;

/// <summary>
/// Static helper for building server-side fingerprints from request headers.
/// </summary>
public static class FingerprintBuilder
{
    /// <summary>
    /// Builds a server-side fingerprint from request headers.
    /// </summary>
    ///
    /// <param name="context">
    /// The HTTP context to build the fingerprint from.
    /// </param>
    ///
    /// <returns>
    /// A 64-character lowercase hex string (SHA-256 hash).
    /// </returns>
    /// <remarks>
    /// The fingerprint is computed from:
    /// SHA-256(User-Agent + "|" + Accept-Language + "|" + Accept-Encoding + "|" + Accept)
    /// This is for logging/analytics only, NOT for rate limiting.
    /// </remarks>
    public static string Build(HttpContext context)
    {
        var headers = context.Request.Headers;

        var userAgent = headers.UserAgent.FirstOrDefault() ?? string.Empty;
        var acceptLanguage = headers.AcceptLanguage.FirstOrDefault() ?? string.Empty;
        var acceptEncoding = headers.AcceptEncoding.FirstOrDefault() ?? string.Empty;
        var accept = headers.Accept.FirstOrDefault() ?? string.Empty;

        var input = $"{userAgent}|{acceptLanguage}|{acceptEncoding}|{accept}";

        return ComputeSha256Hash(input);
    }

    /// <summary>
    /// Computes a SHA-256 hash of the input string.
    /// </summary>
    ///
    /// <param name="input">
    /// The string to hash.
    /// </param>
    ///
    /// <returns>
    /// A 64-character lowercase hex string.
    /// </returns>
    private static string ComputeSha256Hash(string input)
    {
        var bytes = Encoding.UTF8.GetBytes(input);
        var hash = SHA256.HashData(bytes);
        return Convert.ToHexStringLower(hash);
    }
}

File: backends/node/shared/implementations/middleware/request-enrichment/default/src/fingerprint-builder.ts
import { createHash } from "node:crypto";

/**
 * Extracts the first value from a header that may be a string or string array.
 */
function getHeaderValue(
  headers: Record<string, string | string[] | undefined>,
  key: string,
): string {
  const value = headers[key];
  if (value === undefined) return "";
  return Array.isArray(value) ? (value[0] ?? "") : value;
}

/**
 * Builds a combined device fingerprint from all available signals.
 * Formula: SHA-256(clientFingerprint + serverFingerprint + clientIp)
 *
 * Always produces a value — if clientFingerprint is missing, uses empty string
 * (degraded but still unique per server-fingerprint + IP combination).
 *
 * @param clientFingerprint - Client-provided fingerprint (from d2-cfp cookie or header). May be undefined.
 * @param serverFingerprint - Server-computed fingerprint (64-char hex).
 * @param clientIp - Resolved client IP address.
 */
export function buildDeviceFingerprint(
  clientFingerprint: string | undefined,
  serverFingerprint: string,
  clientIp: string,
): string {
  const input = `${clientFingerprint ?? ""}${serverFingerprint}${clientIp}`;
  return createHash("sha256").update(input, "utf8").digest("hex");
}

/**
 * Builds a server-side fingerprint from request headers.
 * Mirrors D2.Shared.RequestEnrichment.Default.FingerprintBuilder.Build in .NET.
 *
 * The fingerprint is SHA-256("{user-agent}|{accept-language}|{accept-encoding}|{accept}"),
 * returned as a 64-character lowercase hex string.
 *
 * @param headers - Plain headers object (framework-agnostic, lowercase keys).
 */
export function buildServerFingerprint(
  headers: Record<string, string | string[] | undefined>,
): string {
  const userAgent = getHeaderValue(headers, "user-agent");
  const acceptLanguage = getHeaderValue(headers, "accept-language");
  const acceptEncoding = getHeaderValue(headers, "accept-encoding");
  const accept = getHeaderValue(headers, "accept");

  const input = `${userAgent}|${acceptLanguage}|${acceptEncoding}|${accept}`;

  return createHash("sha256").update(input, "utf8").digest("hex");
}
